scipy/sparse/linalg/tests/test_propack.py

```diff
@@ -0,0 +1,272 @@
+import os
+import pytest
+import sys
+
+import numpy as np
+from numpy.testing import assert_allclose, assert_raises
+
+from scipy.sparse.linalg._svdp import _svdp
+from scipy.sparse import csr_matrix, csc_matrix, coo_matrix
+
+TOLS = {
+    np.float32: 1e-4,
+    np.float64: 1e-8,
+    np.complex64: 1e-4,
+    np.complex128: 1e-8,
+}
+
+_dtype_map = {
+    "single": np.float32,
+    "double": np.float64,
+    "complex8": np.complex64,
+    "complex16": np.complex128,
+}
+
+
+def is_complex_type(dtype):
+    return np.dtype(dtype).char.isupper()
+
+
+def is_32bit():
+    return sys.maxsize <= 2**32  # (usually 2**31-1 on 32-bit)
+
+
+_dtype_testing = []
+for dtype in _dtype_map:
+    if 'complex' in dtype and is_32bit():
+        # PROPACK has issues w/ complex on 32-bit; see gh-14433
+        marks = [pytest.mark.skip]
+    elif 'complex' in dtype:
+        marks = [pytest.mark.slow]  # type: ignore[list-item]
+    else:
+        marks = []
+    _dtype_testing.append(pytest.param(dtype, marks=marks))
+_dtype_testing = tuple(_dtype_testing)  # type: ignore[assignment]
+
+
+def generate_matrix(constructor, n, m, f,
+                    dtype=float, rseed=0, **kwargs):
+    """Generate a random sparse matrix"""
+    rng = np.random.RandomState(rseed)
+    if is_complex_type(dtype):
+        M = (- 5 + 10 * rng.rand(n, m)
+             - 5j + 10j * rng.rand(n, m)).astype(dtype)
+    else:
+        M = (-5 + 10 * rng.rand(n, m)).astype(dtype)
+    M[M.real > 10 * f - 5] = 0
+    return constructor(M, **kwargs)
+
+
+def assert_orthogonal(u1, u2, rtol, atol):
+    """Check that the first k rows of u1 and u2 are orthogonal"""
+    A = abs(np.dot(u1.conj().T, u2))
+    assert_allclose(A, np.eye(u1.shape[1], u2.shape[1]), rtol=rtol, atol=atol)
+
+
+def check_svdp(n, m, constructor, dtype, k, irl_mode, which, f=0.8):
+    tol = TOLS[dtype]
+
+    M = generate_matrix(np.asarray, n, m, f, dtype)
+    Msp = constructor(M)
+
+    u1, sigma1, vt1 = np.linalg.svd(M, full_matrices=False)
+    u2, sigma2, vt2, _ = _svdp(Msp, k=k, which=which, irl_mode=irl_mode,
+                               tol=tol)
+
+    # check the which
+    if which.upper() == 'SM':
+        u1 = np.roll(u1, k, 1)
+        vt1 = np.roll(vt1, k, 0)
+        sigma1 = np.roll(sigma1, k)
+
+    # check that singular values agree
+    assert_allclose(sigma1[:k], sigma2, rtol=tol, atol=tol)
+
+    # check that singular vectors are orthogonal
+    assert_orthogonal(u1, u2, rtol=tol, atol=tol)
+    assert_orthogonal(vt1.T, vt2.T, rtol=tol, atol=tol)
+
+
+@pytest.mark.parametrize('ctor', (np.array, csr_matrix, csc_matrix))
+@pytest.mark.parametrize('precision', _dtype_testing)
+@pytest.mark.parametrize('irl', (True, False))
+@pytest.mark.parametrize('which', ('LM', 'SM'))
+def test_svdp(ctor, precision, irl, which):
+    np.random.seed(0)
```

Comment:
We should not use global seeding at all (do like in `test_svds`)

Reply:
Has there been a discussion about imposing this requirement on tests? gh-14322 is still in discussion.

When this code is run, there are many functions that use random numbers. Tracking all of these down and passing `rng` through the function calls is inconvenient. The reviewer would also need to verify that all functions that should be receiving the `rng` are getting it. So I don't see why we should do that here** if the issue is still under discussion and there are still many other occurrences of `np.random.seed` in other tests.

** Of course, one reason to do it is because you say so, and you are considering merging this PR : ) If that's the rationale, I will do it!

Reply:
> Has there been a discussion about imposing this requirement on tests? [gh-14322](https://github.com/scipy/scipy/issues/14322) is still in discussion.

Not that I know. What it accepted though is that we should use the `np.random.Generator` for new code.

> When this code is run, there are many functions that use random numbers. Tracking all of these down and passing `rng` through the function calls is inconvenient. The reviewer would also need to verify that all functions that should be receiving the `rng` are getting it. So I don't see why we should do that here** if the issue is still under discussion and there are still many other occurrences of `np.random.seed` in other tests.
 
IMO we should try to avoid using global seeding in new code (testing or not). OC, if this cannot be easily done, because there would be some underlying code we cannot touch which requires that, then sure we keep it.

> ** Of course, one reason to do it is because you say so, and you are considering merging this PR : ) If that's the rationale, I will do it!

I consider merging, but I would not impose things like that 😅

Reply:
Thanks, in that case I'll leave it alone.